Classes/Emogrifier.php

```diff
@@ -329,6 +367,107 @@ protected function process(\DOMDocument $xmlDocument)
+     * node.
+     *
+     * @param string[] $styles    the new CSS styles taken from the global styles
+     *                            to be applied to this node.
+     * @param \DOMNode $node      node to apply styles to.
+     *
+     * @return void
+     */
+    private function mapCssToHtmlAttributes(array $styles, \DOMNode $node)
+    {
+        foreach ($styles as $property => $value) {
+            // Strip !important indicator
+            $value = trim(str_replace('!important', '', $value));
+            $this->mapCssToHtmlAttribute($property, $value, $node);
+        }
+    }
+
+    /**
+     * Tries to apply the CSS style to $node as an attribute.
+     *
+     * This method maps a CSS rule to HTML attributes and adds those to the
+     * node.
+     *
+     * @param string $property    the name of the CSS property to map.
+     * @param string $value       the value of the style rule to map.
+     * @param \DOMNode $node      node to apply styles to.
+     *
+     * @return void
+     */
+    private function mapCssToHtmlAttribute($property, $value, \DOMNode $node)
+    {
+        if (isset($this->cssToHtmlMap[$property])) {
+            $mapping = $this->cssToHtmlMap[$property];
+            if (is_string($mapping)) {
+                // Simple attribute mapping
+                $attribute = $mapping;
+            } else {
+                if (isset($mapping['nodes']) && !in_array($node->nodeName, $mapping['nodes'])) {
+                    // The node is not in the whitelist
+                    return;
+                }
+                if (isset($mapping['values']) && !in_array($value, $mapping['values'])) {
+                    // The value cannot be mapped
+                    return;
+                }
+                $attribute = $mapping['attribute'];
+            }
+            $node->setAttribute($attribute, $value);
+            return;
+        }
+        /** The property is not found in the 'simple-mapping' table. Check the
+         * more exotic cases
+         */
+        switch ($property) {
+            case 'background':
+                // Parse out the color, if any
+                $styles = explode(' ', $value);
+                $first = $styles[0];
+                if (!is_numeric(substr($first, 0, 1)) && substr($first, 0, 3) !== 'url') {
+                    // This is not a position or image, assume it's a color
+                    $node->setAttribute('bgcolor', $first);
+                }
+                break;
+            case 'width':
+            case 'height':
+                // Remove 'px'. This regex only conserves numbers and %
+                $number = preg_replace('/[^0-9.%]/', '', $value);
+                $node->setAttribute($property, $number);
+                break;
+            case 'margin':
+                if ($node->nodeName === 'table' || $node->nodeName === 'img') {
+                    $margins = preg_split('/\s+/', $value);
```

Comment:
Could you please use a double backslash here so it's clear to the reader that you're not escaping for the PHP string, but for the regular expression?